lightning/src/ln/channel.rs

```diff
@@ -3352,44 +3376,125 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 	}
 
-	fn maybe_propose_first_closing_signed<F: Deref>(&mut self, fee_estimator: &F) -> Option<msgs::ClosingSigned>
+	/// Calculates and returns our minimum and maximum closing transaction fee amounts, in whole
+	/// satoshis. The amounts remain consistent unless a peer disconnects/reconnects or we restart,
+	/// at which point they will be recalculated.
+	fn calculate_closing_fee_limits<F: Deref>(&mut self, fee_estimator: &F) -> (u64, u64)
 		where F::Target: FeeEstimator
 	{
-		if !self.is_outbound() || !self.pending_inbound_htlcs.is_empty() || !self.pending_outbound_htlcs.is_empty() ||
-				self.channel_state & (BOTH_SIDES_SHUTDOWN_MASK | ChannelState::AwaitingRemoteRevoke as u32) != BOTH_SIDES_SHUTDOWN_MASK ||
-				self.last_sent_closing_fee.is_some() || self.pending_update_fee.is_some() {
-			return None;
-		}
+		if let Some((min, max)) = self.closing_fee_limits { return (min, max); }
 
+		// Propose a range from our current Background feerate to our Normal feerate plus our
+		// force_close_avoidance_max_fee_satoshis.
+		// If we fail to come to consensus, we'll have to force-close.
 		let mut proposed_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Background);
-		if self.feerate_per_kw > proposed_feerate {
-			proposed_feerate = self.feerate_per_kw;
-		}
-		assert!(self.shutdown_scriptpubkey.is_some());
+		let normal_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Normal);
+		let mut proposed_max_feerate = if self.is_outbound() { normal_feerate } else { u32::max_value() };
+
+		// The spec requires that (when the channel does not have anchors) we only send absolute
+		// channel fees no greater than the absolute channel fee on the current commitment
+		// transaction. It's unclear *which* commitment transaction this refers to, and there isn't
+		// very good reason to apply such a limit in any case. We don't bother doing so, risking
+		// some force-closure by old nodes, but we wanted to close the channel anyway.
+
+		if let Some(target_feerate) = self.target_closing_feerate_sats_per_kw {
+			let min_feerate = if self.is_outbound() { target_feerate } else { cmp::min(self.feerate_per_kw, target_feerate) };
+			proposed_feerate = cmp::max(proposed_feerate, min_feerate);
+			proposed_max_feerate = cmp::max(proposed_max_feerate, min_feerate);
+		}
+
+		// Note that technically we could end up with a lower minimum fee if one sides' balance is
+		// below our dust limit, causing the output to disappear. We don't bother handling this
```

Comment:
As observed yesterday during the meeting, I think enforcing the `dust_limit_satoshis` on closing transactions is a bit dubious as it would re-introduce a negotiation among counterparties on the "honest" `dust_limit_satoshis` at ongoing feerate.

If the output type is known, let's just enforced compared to known Core values and otherwise take the _non-safety_ risk of non-propagating transactions. We can always add checks when new outputs types will be deployed with a corresponding policy. Though spec discussion.

Reply:
Right, I think we'll want to use the "each side can unilaterally remove its output" thing, but its somewhat broken in the current spec. Will work with t-bast on it.